### ISSUE TITLE
Incorrect type evaluation for call that captures a generic constructor by a ParamSpec

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#5666]
```python
from typing import Generic, TypeVar, Callable, Type
from typing_extensions import ParamSpec, Protocol, reveal_type


P = ParamSpec("P")
T = TypeVar("T")
T_co = TypeVar("T_co", covariant=True)


class TestProtocol(Protocol[P, T_co]):
    def __init__(self, *args: P.args, **kwargs: P.kwargs) -> None:
        ...

    def __new__(cls) -> T_co:
        ...


class Test(Generic[T]):
    def __init__(self, val1: T, val2: int) -> None:
        ...


def test2(val1: T, val2: int) -> Test[T]:
    ...


def func(obj: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    ...


def func2(obj: Type[TestProtocol[P, T]], *args: P.args, **kwargs: P.kwargs) -> T:
    ...


val_test = Test(val1=1, val2=2)


reveal_type(Test(val1=1, val2=2))  # Test[int], as expected.
reveal_type(test2(val1=1, val2=2))  # Test[int], as expected.
reveal_type(func(Test, val1=1, val2=2))  # Test[Unknown], however, expecting Test[int].
reveal_type(func2(Test, val1=1, val2=2))  # Test[Unknown], however, expecting Test[int].
```

_Originally posted by @cainmagi in https://github.com/microsoft/pyright/issues/5614#issuecomment-1668580271_

**Comment by erictraut** [comment]
Constructor type evaluation and validation involves significant special-cased code to handle the fact that constructors are composite calls involving a call to the metaclass `__call__` followed by a call to the class's `__new__` method and finally to the class's `__init__` method. These three methods can all have different signatures and reference class-scoped and method-scoped type variables. 

When you assign a class to a variable or a parameter whose type is a `Callable`, the constructor gets "squashed" into a single call that attempts to approximate the signature of the group of methods mentioned above. This "squashing" process eliminates opportunities for handling some of the special casing that's possible with the full constructor.

I investigated the possibility of handling `func(Test, val1=1, val2=2)`. Unfortunately, making this work will break other more common use cases that show up in real-world code, so I decided not to commit it.

The case of `func2(Test, val1=1, val2=2)` will definitely not work as you're hoping for the reasons I mention above. 

You'll need to find an alternative if your solution relies on this pattern.
